src/test/java/com/palantir/common/streams/BackpressureSpliteratorTests.java

```diff
@@ -0,0 +1,134 @@
+        verifyZeroInteractions(consumer);
+    }
+
+    @Test
+    public void throwsIfSuppliedFutureThrows() {
+        CompletableFuture<String> someFuture = new CompletableFuture<>();
+        Spliterator<String> spliterator =
+                BackpressureSpliterator.create(1, Stream.of(someFuture).spliterator());
+        someFuture.completeExceptionally(new RuntimeException());
+        assertThatExceptionOfType(CompletionException.class).isThrownBy(() -> spliterator.tryAdvance(consumer));
+    }
+
+    @Test
+    public void onlyRunsUpToDesiredConcurrencyTasksSimultaneously() {
+        Spliterator<String> spliterator = BackpressureSpliterator.create(1, sourceSpliterator);
+
+        verify(sourceSpliterator, times(1)).tryAdvance(any());
+        String firstData = "firstData";
+        future.complete(firstData);
+        assertThat(spliterator.tryAdvance(consumer)).isTrue();
+        verify(consumer).accept(firstData);
+
+        String secondData = "secondData";
+        otherFuture.complete(secondData);
+        assertThat(spliterator.tryAdvance(consumer)).isTrue();
+        verify(consumer).accept(secondData);
+
+        assertThat(spliterator.tryAdvance(consumer)).isFalse();
+        verifyNoMoreInteractions(consumer);
+    }
+
+    @Test
+    public void runsDesiredConcurrencyTasksSimultaneously() {
+        BackpressureSpliterator.create(
+                2,
+                sourceSpliterator);
+
+        verify(sourceSpliterator, times(2)).tryAdvance(any());
+    }
+
+    @Test
+    public void startNextTaskBeforeConsumingValue() {
+        Spliterator<String> spliterator = BackpressureSpliterator.create(
+                1,
+                sourceSpliterator);
+
+        String data = "data";
+        future.complete(data);
+        otherFuture.complete("some other data");
+
+        // check that we kicked off next task before the consumer has been called
+        doAnswer(inv -> {
+            verify(sourceSpliterator, times(2)).tryAdvance(any());
+            return null;
+        }).when(consumer).accept(data);
+
+        spliterator.tryAdvance(consumer);
+    }
+
+    // This test exists because of an implementation bug while writing this.
+    @Test
+    public void canHandleFutureAlreadyCompleted() {
+        CompletableFuture<String> someFuture = new CompletableFuture<>();
+        String data = "data";
+        someFuture.complete(data);
+
+        Spliterator<String> spliterator = BackpressureSpliterator.create(1, Stream.of(someFuture).spliterator());
+
+        assertThat(spliterator.tryAdvance(consumer)).isTrue();
+        verify(consumer).accept(data);
+        assertThat(spliterator.tryAdvance(consumer)).isFalse();
+    }
+
+    @Test
+    public void testEstimateSize_hasSize() {
+        long estimate = 5L;
+        when(sourceSpliterator.estimateSize()).thenReturn(estimate);
+        assertThat(BackpressureSpliterator.create(2, sourceSpliterator).estimateSize()).isEqualTo(estimate + 2);
```

Comment:
Size should be equal to estimate, not estimate + 2?

Reply:
nope! the correct estimate is the estimate of stuff we currently have + estimate of stuff in our spliterator, right?